pilot/pkg/model/xds_cache.go

```diff
@@ -144,36 +164,96 @@ func newLru() simplelru.LRUCache {
 	return l
 }
 
-func (l *lruCache) Add(entry XdsCacheEntry, value *any.Any) {
+// assertUnchanged checks that a cache entry is not changed. This helps catch bad cache invalidation
+// We should never have a case where we overwrite an existing item with a new change. Instead, when
+// config sources change, Clear/ClearAll should be called. At this point, we may get multiple writes
+// because multiple writers may get cache misses concurrently, but they ought to generate identical
+// configuration. This also checks that our XDS config generation is deterministic, which is a very
+// important property.
+//
+// DO NOT CALL UNLESS EnableUnsafeAssertions IS ENABLED
+func assertUnchanged(existing *any.Any, replacement *any.Any) {
```

Comment:
Instead of textual comments, consider using build tag for assertions, and supply two implementations of this interface with release impl being a noop.

Reply:
I was thinking about this as well, since its how its done in the C world. Some annoyances are that we need a 2 completely distinct file to define the function in both with the build tags - its a bit of a pain. Additionally, since each assertion is context sensitive, we either need to have a bunch of these split files or have one big context-sensitive assertion file. Or we have a generic `assert` function, but then it needs to take a function and usage is a bit awkward. And now we need to deal with a completely separate binary rather than a simple env var.

I do share your concern about comments though - I think it can be addressed by adding the `if EnableUnsafeAssertions` inside the function though so it cannot be misused?